### PR TITLE
JAMES-2285 MailQueueContract only

### PR DESCRIPTION
**Pull request by chibenwa**
First subpart of the infamous https://github.com/linagora/james-project/pull/1222

### CODE REVIEW

File: server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/LegacyJMSMailQueueTest.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.util.Arrays;
+
+import org.apache.activemq.broker.BrokerService;
+import org.apache.activemq.broker.region.policy.PolicyEntry;
+import org.apache.activemq.broker.region.policy.PolicyMap;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+
+public class LegacyJMSMailQueueTest extends AbstractJMSMailQueueTest {
```
Comment (rouazana) [inline server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/LegacyJMSMailQueueTest.java]: Is there any reason to keep it at the end of this PR?

Comment (chibenwa) [reply server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/LegacyJMSMailQueueTest.java]: Yes. It is needed for the ActiveMQ test class. It is resolved here: https://github.com/linagora/james-project/pull/1228

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/NotifyMailetsMessageTest.java [hunk 2]
```diff
@@ -195,6 +195,7 @@ public void generateMessageShouldSpecifySizeInAReadableWay() throws Exception {
         assertThat(generateMessage).contains("Size: 5.9 KiB");
     }
 
+    @SuppressWarnings("unchecked")
```
Comment (mbaechler) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/NotifyMailetsMessageTest.java]: why ?

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/ExecutorExtension.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+
+public class ExecutorExtension implements ParameterResolver, BeforeEachCallback, AfterEachCallback {
+
+    private ExecutorService executorService;
+
+    @Override
+    public void beforeEach(ExtensionContext context) throws Exception {
+        executorService = Executors.newFixedThreadPool(2);
```
Comment (mbaechler) [inline server/queue/queue-api/src/test/java/org/apache/james/queue/api/ExecutorExtension.java]: maybe you should not limit the number of thread for an extension

File: mailet/base/src/test/java/org/apache/mailet/base/test/MimeMessageBuilder.java [hunk 4]
```diff
@@ -187,18 +186,6 @@ public BodyPart build() throws IOException, MessagingException {
         }
     }
 
-    public static MimeMessage defaultMimeMessage() {
```
Comment (mbaechler) [inline mailet/base/src/test/java/org/apache/mailet/base/test/MimeMessageBuilder.java]: these methods actually makes sense in a builder as they are factory methods

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+        MailQueue.MailQueueItem mailQueueItem = getMailQueue().deQueue();
+        assertThat(mailQueueItem.getMail().getLastUpdated())
+            .isEqualTo(lastUpdated);
+    }
+
+    @Test
+    default void queueShouldPreserveName() throws Exception {
+        getMailQueue().enQueue(defaultMail()
+            .name("name")
```
Comment (mbaechler) [inline server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java]: extract to variable ?

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+        MailQueue.MailQueueItem mailQueueItem = getMailQueue().deQueue();
+        assertThat(mailQueueItem.getMail().getPerRecipientSpecificHeaders()
+            .getHeadersForRecipient(RECIPIENT1))
+            .containsOnly(header);
+    }
+
+    @Test
+    default void dequeueShouldBeFifo() throws Exception {
+        getMailQueue().enQueue(defaultMail()
+            .name("name1")
```
Comment (mbaechler) [inline server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java]: extract to var ?

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        MailQueue.MailQueueItem mailQueueItem1 = getMailQueue().deQueue();
+        mailQueueItem1.done(true);
+        MailQueue.MailQueueItem mailQueueItem2 = getMailQueue().deQueue();
+        mailQueueItem2.done(true);
+        assertThat(mailQueueItem1.getMail().getName()).isEqualTo("name1");
+        assertThat(mailQueueItem2.getMail().getName()).isEqualTo("name2");
+    }
+
+    @Test
+    default void dequeueCouldBeInterleaving() throws Exception {
```
Comment (mbaechler) [inline server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java]: what is supposed to be "invertleaving" ?

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+            .build());
+        getMailQueue().enQueue(defaultMail()
+            .name("name2")
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem1 = getMailQueue().deQueue();
+        MailQueue.MailQueueItem mailQueueItem2 = getMailQueue().deQueue();
+        mailQueueItem2.done(true);
+        mailQueueItem1.done(true);
+        assertThat(mailQueueItem1.getMail().getName()).isEqualTo("name1");
```
Comment (mbaechler) [inline server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java]: I don't understand why you use this assertion to check the state of your queue. What does it tell you about the result of `done`calls ?

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+
+        getMailQueue().deQueue();
+
+        Future<?> future = executorService.submit(Throwing.runnable(() -> getMailQueue().deQueue()));
+        assertThatThrownBy(() -> future.get(2, TimeUnit.SECONDS))
+            .isInstanceOf(TimeoutException.class);
+    }
+
+    @Test
+    default void deQueueShouldFreezeWhenNoMail(ExecutorService executorService) throws Exception {
```
Comment (mbaechler) [inline server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java]: It's usually called "block" not "freeze"

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+            .isInstanceOf(TimeoutException.class);
+    }
+
+    @Test
+    default void deQueueShouldWaitForAMailToBeEnqueued(ExecutorService executorService) throws Exception {
+        Mail mail = defaultMail()
+            .name("name")
+            .build();
+        Future<MailQueue.MailQueueItem> tryDequeue = executorService.submit(() -> getMailQueue().deQueue());
+        executorService.submit(Throwing.runnable(() -> getMailQueue().enQueue(mail)));
```
Comment (mbaechler) [inline server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java]: why do you need this one to be run asynchronously ?

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+
+        assertThat(tryDequeue.get().getMail().getName()).isEqualTo("name");
+    }
+
```
Comment (mbaechler) [inline server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java]: Other tests cases ideas :
are failed transactions pushed back as next element to be dequeued or elsewhere ?
what happens when you call `done` in various broken order with various values ?
(not sure it doesn't exist yet) do you need to ack pending elements to consume new ones ?

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueFixture.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+import java.util.Date;
+
+import javax.mail.MessagingException;
+import javax.mail.internet.MimeMessage;
+
+import org.apache.mailet.base.test.FakeMail;
+import org.apache.mailet.base.test.MimeMessageBuilder;
+
+public interface MailQueueFixture {
```
Comment (mbaechler) [inline server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueFixture.java]: A Fixture is supposed to be about built objects, not util methods.

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/NotifyMailetsMessageTest.java (hunk 2 above)
Comment (chibenwa) [reply server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/NotifyMailetsMessageTest.java]: @aduprat  ?

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java (hunk 5 above)
Comment (chibenwa) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java]: When I do extract I'm told "that's readable"

When I do not extract, then people ask me this.

A bit depressing. I'll do it only if you really urge me for it.

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java (hunk 6 above)
Comment (chibenwa) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java]: item

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java (hunk 7 above)
Comment (chibenwa) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java]: Maybe "dequeueCanBeChainedBeforeAck"

Comment (mbaechler) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java]: great

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueFixture.java (hunk 12 above)
Comment (chibenwa) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueFixture.java]: builder is statefull, so having it as a field won't work.

Would you consider cleaner to have yet another separated MailQueueTestUtils class? Me no.

Comment (mbaechler) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueFixture.java]: Are you arguing that using Fixture word for the wrong use case is not a problem ? Why would we choose this word if it does not talk about util methods ?

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java (hunk 8 above)
Comment (chibenwa) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java]: Sorry I just ensure the dequeued mail identity... What would you expect?

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/NotifyMailetsMessageTest.java (hunk 2 above)
Comment (aduprat) [reply server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/NotifyMailetsMessageTest.java]: It's on the _when_ line:
```
when(mail.getMessageSize()).thenThrow(MessagingException.class);
```
the warning is:
```
Type safety: A generic array of Class<? extends Throwable> is created for a varargs parameter
```

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/ExecutorExtension.java (hunk 3 above)
Comment (chibenwa) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/ExecutorExtension.java]: What exact threadPool would you expect then?

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java (hunk 11 above)
Comment (chibenwa) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java]:
```
what happens when you call done in various broken order with various values ?
```

Sorry, I don't get this one?

Comment (chibenwa) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java]:
```
do you need to ack pending elements to consume new ones ?
```

That's what `dequeueCouldBeInterleaving` demonstrate.

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/ExecutorExtension.java (hunk 3 above)
Comment (mbaechler) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/ExecutorExtension.java]: newCachedThreadPool ?

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueFixture.java (hunk 12 above)
Comment (chibenwa) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueFixture.java]: What wording would you propose?

Comment (mbaechler) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueFixture.java]: `Mails` ?

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/ExecutorExtension.java (hunk 3 above)
Comment (chibenwa) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/ExecutorExtension.java]: As you wish

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/NotifyMailetsMessageTest.java (hunk 2 above)
Comment (chibenwa) [reply server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/NotifyMailetsMessageTest.java]: Ok let's keep it. Would that be Ok matthieu?

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java (hunk 8 above)
Comment (mbaechler) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java]: Ok so, let's imagine the queue doesn't support calling done in the wrong order and decide to re-enqueue mail1 when you call mail2.done(true). Your assertion doesn't checks that.
You should check that the queue is empty, for example.

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java (hunk 11 above)
Comment (mbaechler) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java]: > what happens when you call done in various broken order with various values ?

* queue 3 messages, dequeue the 3, call done(false) on 3, then 2, then 1. Check message order in the queue.

* queue 3 messages, dequeue the 3, call done(true) on 3, then done(false) on 2 and 1. Check message order in the queue.

etc.

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueFixture.java (hunk 12 above)
Comment (chibenwa) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueFixture.java]: Sorry I don't understand.

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/NotifyMailetsMessageTest.java (hunk 2 above)
Comment (mbaechler) [reply server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/NotifyMailetsMessageTest.java]: no that's not ok, please use : 
```
        when(mail.getMessageSize()).thenThrow(new MessagingException());
```

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueFixture.java (hunk 12 above)
Comment (mbaechler) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueFixture.java]: I suggest the class to be called `Mails` like for example `Lists` help building List in guava.

File: mailet/base/src/test/java/org/apache/mailet/base/test/MimeMessageBuilder.java (hunk 4 above)
Comment (aduprat) [reply mailet/base/src/test/java/org/apache/mailet/base/test/MimeMessageBuilder.java]: No, I do not agree, they do not depend on _MimeMessageBuilder_, they can be thought as helpers.

**Comment by aduprat** [comment]
test this please

**Comment by chibenwa** [comment]
```
CassandraSharingTest » OutOfMemory Java heap space
``` 

test this please

**Comment by chibenwa** [comment]
```
CassandraSharingTest » OutOfMemory Java heap space
```

test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

Java heap space

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**